DESIGN.md

```diff
@@ -0,0 +1,400 @@
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {
+    Name     string
+    Value    int
+    HashFunc crypto.Hash
+}
+
+func I2OSP(x *big.Int, n int) ([]byte, error)
+func OS2IP(x []byte) *big.Int
+
+type ProtectedHeader map[interface{}]interface{}
+    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error
+    func (h *ProtectedHeader) SetAlgorithm(alg *Algorithm)
+
+type UnprotectedHeader map[interface{}]interface{}
+
+type Headers struct {
+    RawProtected   []byte
+    RawUnprotected []byte
+    Protected      ProtectedHeader
+    Unprotected    UnprotectedHeader
+}
+
+type Signature struct {
+    Headers   Headers
+    Signature []byte
+}
+    func NewSignature() *Signature
+    func (s *Signature) MarshalCBOR() ([]byte, error)
+    func (s *Signature) UnmarshalCBOR(data []byte) error
+
+type SignMessage struct {
+    Headers    Headers
+    Payload    []byte
+    Signatures []*Signature
+}
+    func NewSignMessage() *SignMessage
+    func ParseSignMessage(data []byte) (*SignMessage, error)
+    func (m *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) UnmarshalCBOR(data []byte) error
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) error
+    func (m *SignMessage) Verify(external []byte, verifier []Verifier) error
+
+type Sign1Message struct {
+    Headers   Headers
+    Payload   []byte
+    Signature []byte
+}
+    func NewSign1Message() *Sign1Message
+    func ParseSign1Message(data []byte) (*Sign1Message, error)
+    func (m *Sign1Message) MarshalCBOR() ([]byte, error)
+    func (m *Sign1Message) UnmarshalCBOR(data []byte) error
+    func (m *Sign1Message) Sign(rand io.Reader, external []byte, signer Signer) error
+    func (m *Sign1Message) Verify(external []byte, verifier Verifier) error
+
+type Signer interface {
+    Algorithm() *Algorithm

```

Comment:
Why do we have to bind a signer to an algorithm? This precludes using the same signer for different algorithms, i.e. reusing the same key ECDSA P-256 key with a 256 and a 512 hash.

Same applies for the Verifier.

Reply:
A new signer instance should always be created for a new algorithm, even for the same key.

Key reuse requires extraordinary care in terms of cryptography, including intensive cryptoanalysis or even math proofs. Improper reuse of keys can make the entire scheme insecure.

Reply:
Got it. 

I'm still not fully convinced about `Signer` design, as it's mixing a behavior, `Sign()`, with am orthogonal property, `Algorithm()`. Changing the algorithm does not affect Sign's behavior so, IMO, both concepts shouldn't be grouped in the same interface, unless strictly necessary for safety or design limitations.

Another major drawback is that to fulfill the Signer interface one must import the `go-cose` module: `Algorithm()` returns a go-cose type. This goes against Go's interface design principle of decoupling dependencies. 

Here is an alternate draft design:

```go
type Signer interface {
    Sign(rand io.Reader, digest []byte) ([]byte, error)
}

type Signature struct {
    Signer
    Algorithm Algorithm
    Headers Headers
}

func NewSignature(alg *Algorithm, headers Headers, key crypto.Signer) (Signature, error)

func Sign(rand io.Reader, headers Headers, payload, external []byte, signers ...Signature) (*SignMessage, error)
func Sign1(rand io.Reader, payload, external []byte, signer Signature) (*SignMessage, error)
```

With this design we kill two birds with one stone:

- Signer interface is smaller and do not expect go-cose types, therefore easier to compose and implement.
- A Signer can't be used without specifying an algorithm when signing messages.
- The Signature struct implements the Signer behavior and adds additional signing properties, which would work well with the `Sign` functions proposed in https://github.com/veraison/go-cose/pull/14/files#r836487473.

Reply:
I retract this proposal. The algorithm does affect the behavior of the signer, so it makes sense to keep it in the same interface. For example, the same RSA private key will use different SHA sizes depending on if the algorithm is PS256, PS384 or PS512.